### PR TITLE
Unit Tests for LocationPickerActivity

### PR DESCRIPTION
**Pull request by Ayan-10**
**Description (required)**

Added unit tests for LocationPickerActivity

### CODE REVIEW

**Comment by madhurgupta10** [comment]
> ```
> fr.free.nrw.commons.navtab.MoreBottomSheetLoggedOutFragmentUnitTests > testOnFeedbackClicked FAILED
>     java.lang.IllegalStateException at MoreBottomSheetLoggedOutFragmentUnitTests.kt:66
> ```
> 
> @madhurgupta10 I think this build failure doesn't belong to this PR.

@Ayan-10 Yes, I have triggered the job again. This is a known issue, not sure what causes it but sometimes this test fails.

File: app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Test
+    @Throws(Exception::class)
+    fun checkActivityNotNull() {
+        Assert.assertNotNull(activity)
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun testBindListeners() {
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt:69]: Add verification to this tests for `mapboxMap`

File: app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        )
+        method.isAccessible = true
+        method.invoke(activity)
+        verify(mapboxMap, times(1))
+            .moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun testOnChanged() {
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt:102]: Add verification to this tests for mapboxMap

File: app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        val method: Method = LocationPickerActivity::class.java.getDeclaredMethod(
+            "bindListeners"
+        )
+        method.isAccessible = true
+        method.invoke(activity)
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun testOnMapReady() {
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/locationpicker/LocationPickerActivityUnitTests.kt:79]: You can refactor your lambda block inside `onMapReady`  to a new function and write additional tests to cover those lines

**Review by madhurgupta10 — CHANGES_REQUESTED** [review]

**Comment by madhurgupta10** [comment]
@Ayan-10 also, please sync your branch with the latest master to compare with the most recent code coverage.

**Comment by madhurgupta10** [comment]
> ```
> droppedMarkerLayer.setProperties(
>     visibility: none
> );
> -> at com.mapbox.mapboxsdk.style.layers.Layer.setProperties(Layer.java:47)
> Actual invocation has different arguments:
> droppedMarkerLayer.setProperties(
>     visibility: none
> );
> ```
> 
> Getting this error while verifying `droppedMarkerLayer`. Although the actual and wanted arguments are the same as you can see in the error.
> 
> The if part of onStyleLoaded is showing `Thread is not initialized ` error on `GeoJsonSource("dropped-marker-source-id")`.

Different objects are compared even though they carry same value. Try `verify(yourMock).yourMethod(any())

**Comment by Ayan-10** [comment]
`droppedMarkerLayer` is verified now.

Thanks, @madhurgupta10

**Comment by madhurgupta10** [comment]
@Ayan-10 `24.69%` coverage for Location Picker Activity is decent but try to get it around 80+

**Comment by Ayan-10** [comment]
@madhurgupta10 I tried my best to cover more lines. But hardly 20-25 more lines can be covered if I refactor some lines from `onCreate` and some methods but 80+ will be possible and the reason is static and final methods/members. Mocikto is behaving unexpectedly while mocking static/final members compared to others. Most of the lines are uncovered because of this. Also, I was doing `whenever(activity.findViewById(R.id.string))` but it's not triggering the `findViewById` in LocationPickerActivity.

**Review by madhurgupta10 — APPROVED** [review]
I am happy to merge these tests, you can try with Instrumentation tests after https://github.com/commons-app/apps-android-commons/pull/4832 is merged to reach the 80% target

**Comment by madhurgupta10** [comment]
@Ayan-10 These methods would be easy to test using mockito
image

**Comment by Ayan-10** [comment]
@madhurgupta10 I tried to test those before but got `Caused by: java.lang.IllegalStateException: No activity` error I thought the error is occurring because of not calling `.create` while creating the activity.

**Comment by madhurgupta10** [comment]
> @madhurgupta10 I tried to test those before but got `Caused by: java.lang.IllegalStateException: No activity` error I thought the error is occurring because of not calling `.create` while creating the activity.

You can update the activity to an instance of MainActivity during those tests.

**Comment by Ayan-10** [comment]
> You can update the activity to an instance of MainActivity during those tests.

If I update the activity as a MainActivity during those tests, then I think methods of MainActivity will be covered instead of LocationPickerActivity. Isn't it?


I was thinking the same error will occur for every method but all errors of those methods are not the same and `onLowMemory` can be tested. **So I added the test for `onLowMemory`** and also attached all the errors.

**onDestroy, onSaveInstanceState and onStart**
```
Caused by: java.lang.IllegalStateException: FragmentManager has been destroyed
	at androidx.fragment.app.FragmentManager.ensureExecReady(FragmentManager.java:1781)
	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1841)
	at androidx.fragment.app.FragmentManager.dispatchDestroy(FragmentManager.java:2599)
	at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:330)
	at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:365)
	at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:233)
	at fr.free.nrw.commons.theme.BaseActivity.onDestroy(BaseActivity.java:50)
	at fr.free.nrw.commons.LocationPicker.LocationPickerActivity.onDestroy(LocationPickerActivity.java:441)
```

**onStop and onPause**
```
Caused by: java.lang.IllegalStateException: No activity
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1484)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2617)
	at androidx.fragment.app.FragmentManager.dispatchPause(FragmentManager.java:2585)
	at androidx.fragment.app.FragmentController.dispatchPause(FragmentController.java:280)
	at androidx.fragment.app.FragmentActivity.onPause(FragmentActivity.java:419)
	at fr.free.nrw.commons.LocationPicker.LocationPickerActivity.onPause(LocationPickerActivity.java:423)

```
**onResume**
```
Caused by: java.lang.NullPointerException
	at fr.free.nrw.commons.theme.BaseActivity.onResume(BaseActivity.java:41)
	at fr.free.nrw.commons.LocationPicker.LocationPickerActivity.onResume(LocationPickerActivity.java:417)

```

**Comment by madhurgupta10** [comment]
Correct, maybe then instrumentation tests is the only option here.

**Pull request merged**